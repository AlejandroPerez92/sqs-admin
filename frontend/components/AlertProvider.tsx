import React, { createContext, useContext, useState, ReactNode } from "react";
import Alert from "./Alert";
import { createPortal } from "react-dom";

type Severity = "error" | "warning" | "info" | "success";

interface AlertContextProps {
  showAlert: (message: string, severity: Severity) => void;
}

interface Alert {
  message: string;
  severity: Severity;
}

const AlertContext = createContext<AlertContextProps | undefined>(undefined);

export const AlertProvider: React.FC<{ children: ReactNode }> = ({
                                                                   children
                                                                 }) => {
  const [alerts, setAlerts] = useState<Map<string, Alert>>(new Map());

  const addAlert = (alert: Alert): void => {
    setAlerts(prevAlerts => {
      const newMap = new Map(prevAlerts);
      const key = Math.random().toString();
      newMap.set(key, alert);
      setTimeout(() => deleteAlert(key.toString()), 3000);

      return newMap;
    });
  };

  const deleteAlert = (key: string): void => {
    setAlerts(prevAlerts => {
      const newMap = new Map(prevAlerts);
      newMap.delete(key);
      return newMap;
    });
  };

  const showAlert = (message: string, severity: Severity) => {
    addAlert({ message, severity });
  };

  return (
    <AlertContext.Provider value={{ showAlert }}>
      {children}
      {createPortal(
        <div
          style={{
            position: "fixed",
            top: 20,
            left: "50%",
            transform: "translateX(-50%)",
            zIndex: 9999,
            width: "90%",
            maxWidth: 500,
            pointerEvents: "none",
          }}
        >
          {Array.from(alerts.entries()).map(([key, alert]) => (
          <Alert
            message={alert.message}
            severity={alert.severity}
            onClose={() => deleteAlert(key)}
            key={key}
          />
          ))}
        </div>,
        document.body
      )}
    </AlertContext.Provider>
  );
};

export const useAlert = (): AlertContextProps => {
  const context = useContext(AlertContext);
  if (!context) {
    throw new Error("useAlert must be used inside AlterProvider wrapper");
  }
  return context;
};
