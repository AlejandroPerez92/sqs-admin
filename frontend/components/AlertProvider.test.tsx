import { render } from "@testing-library/react";
import { AlertProvider, useAlert } from "./AlertProvider";
import { useEffect, useState } from "react";

const TestComponent = () => {
  const { showAlert } = useAlert();
  const [hasAlertBeenShown, setHasAlertBeenShown] = useState(false);

  useEffect(() => {
    if (!hasAlertBeenShown) {
      showAlert("Test error", "error");
      showAlert("Test success", "success");
      setHasAlertBeenShown(true);
    }
  }, [hasAlertBeenShown, showAlert]);

  return <h1>Test</h1>;
};

describe("<AlertProvider /> spec", () => {
  it("renders the Alert and triggers showAlert", () => {
    const view = render(
      <AlertProvider>
        <TestComponent />
      </AlertProvider>
    );

    expect(view).toMatchSnapshot();
  });
});