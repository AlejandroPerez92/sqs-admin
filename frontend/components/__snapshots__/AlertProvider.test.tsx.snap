// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertProvider /> spec renders the Alert and triggers showAlert 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Test
      </h1>
    </div>
    <div
      style="position: fixed; top: 20px; left: 50%; transform: translateX(-50%); z-index: 9999; width: 90%; max-width: 500px; pointer-events: none;"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorError MuiAlert-standardError MuiAlert-standard css-1ex1cfp-MuiPaper-root-MuiAlert-root"
        role="alert"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiAlert-icon css-vab54s-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
            data-testid="ErrorOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-zioonp-MuiAlert-message"
        >
          Test error
        </div>
        <div
          class="MuiAlert-action css-rgppqo-MuiAlert-action"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-4ns38t-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorSuccess MuiAlert-standardSuccess MuiAlert-standard css-1nnotcy-MuiPaper-root-MuiAlert-root"
        role="alert"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiAlert-icon css-vab54s-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
            data-testid="SuccessOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-zioonp-MuiAlert-message"
        >
          Test success
        </div>
        <div
          class="MuiAlert-action css-rgppqo-MuiAlert-action"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-4ns38t-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1>
      Test
    </h1>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
